scheds/rust/scx_layered/src/bpf/main.bpf.c

```diff
@@ -703,15 +650,37 @@ s32 pick_idle_cpu(struct task_struct *p, s32 prev_cpu,
 	 * the preemption attempt fails.
 	 */
 	maybe_refresh_layered_cpus(p, tctx, layer_cpumask, cpus_seq);
-	if (!(cpumask = cast_mask(tctx->layered_mask)))
+	if (!(layered_cpumask = cast_mask(tctx->layered_mask)))
 		return -1;
-	if (from_selcpu && should_try_preempt_first(prev_cpu, layer, cpumask)) {
+	if (from_selcpu && should_try_preempt_first(prev_cpu, layer, layered_cpumask)) {
 		cctx->try_preempt_first = true;
 		return -1;
 	}
 
-	if (!(prev_cctx = lookup_cpu_ctx(prev_cpu)) ||
-	    !(idle_smtmask = scx_bpf_get_idle_smtmask()))
+	/*
+	 * When a layer stays saturated, there's no point in repeatedly
+	 * searching for an idle CPU at different levels. Short-circuit by
+	 * testing whether there are any eligible CPUs first.
+	 */
+	if (READ_ONCE(layer->check_no_idle)) {
+		bool has_idle;
+
+		cpumask = scx_bpf_get_idle_cpumask();
+
+		if (layer->kind == LAYER_KIND_CONFINED)
+			has_idle = bpf_cpumask_intersects(layered_cpumask, cpumask);
+		else
+			has_idle = bpf_cpumask_intersects(p->cpus_ptr, cpumask);
+
+		scx_bpf_put_idle_cpumask(cpumask);
+		if (!has_idle)
+			return -1;
+	}
+
+	if ((nr_llcs > 1 || nr_nodes > 1) &&
+	    !(prev_cctx = lookup_cpu_ctx(prev_cpu)))
+		return -1;
+	if (!(idle_smtmask = scx_bpf_get_idle_smtmask()))
```

Comment:
If a layer sets [`idle_smt`](https://github.com/sched-ext/scx/blob/main/scheds/rust/scx_layered/src/bpf/intf.h#L196) should this set `idle_smtmask` to `scx_bpf_get_idle_cpumask`? I think that's what it used to do, which is a little hacky of overriding `idle_smtmask`. Otherwise it may skip idle smt cpus in [`pick_idle_cpu_from`](https://github.com/sched-ext/scx/blob/main/scheds/rust/scx_layered/src/bpf/main.bpf.c#L593).

Reply:
The mask isn't used at all by `pick_idle_cpu_from()` when `pref_idle_smt` is not set. I don't know why the mask itself needs to be overridden too.

Reply:
> I don't know why the mask itself needs to be overridden too.

Guh... I'm blind, I see it's using `cand`. I've always gotten confused by that function. Sorry for the noise!